### ISSUE TITLE
Client_credential should automatically cache N tokens (<500MB worth of tokens). Some form of eviction must be used and O(1) cache lookup is required.

### ISSUE THREAD

**Issue by bgavrilMS** [AzureAD/microsoft-authentication-library-for-dotnet#4404]
- [ ] An opt-out from this behavior must be provided, e.g. via the CacheOptions API.
- [ ] Approx size or count eviction limit should be configurable.
- [ ] Discuss having a static cache for CCA ROPC, Web App and Web Api

This is a breaking change because some applications rely on CCA object not caching objects

**Comment by SamarthMayya** [comment]
Hello, is this in any way related to the client credential flow of MSAL not being optimized to lookup the in-memory cache in O(1) time?

I saw that in [ITokenCacheInternal](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs#L499), there is a method ```FilterTokensByScope```, which filters the tokens by checking if the scopes in the `requestParameters.Scopes` is contained in the  `item.Scopes`. This is O(N) lookup, if I'm not wrong. 

The Accessor property for in-memory cache has a ConcurrentDictionary that can be used to lookup the cache per scope in O(1): [link](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/InMemoryPartitionedAppTokenCacheAccessor.cs#L25)

This dictionary has two levels, one is a key that has `ClientId_TenantId_AppTokenCache` as the first key, and the second level key has the scope in it. Is there a way we can change the first key to include the scope in it, so that when we get all the access tokens, we only get the tokens cached for a particular scope?


Mainly, is there a way we can change the partition key we [use](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs#L457C19-L457C19)
 to include the scope?

**Comment by bgavrilMS** [comment]
I agree that the lookup is not O(1), but is proportional to the number of resources an app accesses. But in our experience, this number is low, the most I have seen is ~10 resources.

We can change the logic to use the scope, however we will need to have special lookup logic for client credential flow VS user flows. In user flows, if you request "User.Read", you can get "User.Read" and "User.Write" from the STS, so we can't use scopes as key there.
